### DISCUSSION TITLE
If a field changed from disabled to enabled, it passes undefined to validate function.

### DISCUSSION THREAD

**Discussion by mjza** [react-hook-form/react-hook-form#5277] [Q&A]
I want to have a sort of switch between review and edit mode in my form, then I used `disabled` property. 

If this property is controlled with a property in `state`, and its value set to `true` in the beginning, then the value that is passed to the `validate` function would be `undefined`. 

Please look at the following sandbox:

https://codesandbox.io/s/react-hook-form-controlled-input-failed-validation-bug-forked-gxbh5?file=/src/index.js:1749-1764

The input field has value but when we press on submit we will see error!

In my real app, user press on an edit button, the field would change from disabled to enabled, however, when user presses on submit button they will see error while the item has value!

In version 6 it was working, upgrading to version 7 caused this error.

**Comment by mjza** [comment]
Check the console for better understanding:


> Warning: A component is changing a controlled input to be uncontrolled. This is likely caused by the value changing from a defined to undefined, which should not happen. Decide between using a controlled or uncontrolled input element for the lifetime of the component. More info: https://reactjs.org/link/controlled-components
>     at input
>     at form
>     at Login (https://gxbh5.csb.app/src/index.js:48:34)
>     at App (https://gxbh5.csb.app/src/index.js:142:41)

**Comment by mjza** [comment]
Also if the field has the `autoComplete` property, again it does not recognise the value, and will pass `undefined` to the `validate` function, for example:

```jsx
   <input
          name="username"
          placeholder={t("auth:placeholder.email")}
          className={errors.username ? "error" : ""}
          {...register("username", {
                required: true,
                pattern: Helper.emailRegex(true, true),
           })}
          autoComplete="username"
          onChange={this.onChangeEmail}
     />
```

**Comment by leapful** [comment]
@mjza 

This is not a bug. Disabled field value will not be included in form following the web standard. 

Referrer:
https://react-hook-form.com/api/useform/handlesubmit/#main
https://stackoverflow.com/questions/1355728/values-of-disabled-inputs-will-not-be-submitted

Your second issue of attribute "autocomplete" cannot be re-produced too. Plus, the block code in your comment is not using "validate".

**Comment by mjza** [comment]
@leapful I added a new button to my example sandbox. Check it again. 

https://codesandbox.io/s/react-hook-form-controlled-input-failed-validation-bug-forked-gxbh5?file=/src/index.js

Now, you must press on edit button first, then press on submit to see the error.

Unfortunately you didn't read my issue completely, now I added the EDIT button to make it more clear for you.

**Comment by mjza** [comment]
@bluebill1049 If it is not a bug can you tell me why it is working in version 6 and not in version 7?

**Reply by bluebill1049** [reply]
You are mixed uncontrolled with controlled, the moment you see `onChange` and `value`, you should know it's a controlled component and you should use `useController` or `Controller` instead

<img width="680" alt="Screen Shot 2021-05-23 at 4 16 31 pm" src="https://user-images.githubusercontent.com/10513364/119250266-3ddf9d00-bbe2-11eb-893f-8ccbb867d887.png">

here is a working version: 
https://codesandbox.io/s/ancient-snow-ligz6?file=/src/App.js

**Reply by mjza** [reply]
@bluebill1049 
I changed the `value` property to `defaultValue` and it does not work even!
**The problem is, when it is rendering and the field is disabled, and then is changed to enabled, it does not send the value to validator function.** 
In you example in the beginning the field is not disabled!

**Reply by mjza** [reply]
@bluebill1049 I must insist once more, that the same strategy for mixing controlled and uncontrolled was working in version 6 of the library!

**Reply by mjza** [reply]
@bluebill1049 Please take a look here:
https://codesandbox.io/s/cool-fast-wmb02

![](https://i.stack.imgur.com/AJmIf.jpg)

I changed the initial value of `disabled` to `true`.
By pressing on `toggle` it is changed to activate. 
And you can see in the console the result of pressing on submit.

**Reply by mjza** [reply]
@bluebill1049 Have you tried my sandbox?
https://codesandbox.io/s/cool-fast-wmb02

**Reply by mjza** [reply]
@bluebill1049 Sorry I haven't heard you for the last 2 days. Have you seen the sandbox?
Is there any solution for this bug?
I really need to find a solution for this problem soon.

**Comment by leapful** [comment]
@mjza @bluebill1049 

Sorry for the late response, I've bean busy for a while.

The root cause of this issue is from the way react-hook-form handles "disabled" state with "register". With initial disabled state, default value is not set to internal field value of react-hook-form therefore when it switched to "enabled", that value is still not available.

I've submit a pull request at https://github.com/react-hook-form/react-hook-form/pull/5350/files.

**Reply by mjza** [reply]
Thanks for the update. 
I hope this pull request be applied in the next version. 

@bluebill1049 I have a criticism also. I see you always very fast remove the issues to discussion part. It is now clear it was a bug. Please give some time to people to discuss on issue, and if it is 100% clear that an issue is not bug then remove it to discussion or whatever. It is a little bit unpleasant feeling when you remove an issue, that I have experienced for the second time. Maybe not a bad idea to use flags or tags!

**Reply by mjza** [reply]
@leapful Is there any news regarding the issue?

It seems @bluebill1049 hasn't reacted to your pull request!

**Comment by mjza** [comment]
I confirm in version 7.8.4 it works again.

https://codesandbox.io/s/cool-fast-wmb02
